libs/core/src/bloom_filter.cpp

```diff
@@ -0,0 +1,375 @@
+
+namespace fetch {
+namespace internal {
+
+HashSourceFactory::HashSourceFactory(HashSourceFactory::Functions hash_functions)
+  : hash_functions_(std::move(hash_functions))
+{}
+
+HashSourceFactory::~HashSourceFactory() = default;
+
+HashSource HashSourceFactory::operator()(Bytes const &element) const
+{
+  return HashSource(this->hash_functions_, element);
+}
+
+HashSource::HashSource(HashSourceFactory::Functions const &hash_functions,
+                       HashSourceFactory::Bytes const &    input)
+  : data_{}
+{
+  // TODO(WK): lazily evaluate hashes
+  for (auto const &fn : hash_functions)
+  {
+    auto const hashes = fn(input);
+    data_.insert(data_.cend(), hashes.cbegin(), hashes.cend());
+  }
+}
+
+HashSource::~HashSource()                      = default;
+HashSource::HashSource(HashSource &&) noexcept = default;
+HashSource &HashSource::operator=(HashSource &&) noexcept = default;
+
+HashSource::HashSourceIterator HashSource::begin() const
+{
+  return HashSource::HashSourceIterator{this, 0u};
+}
+
+HashSource::HashSourceIterator HashSource::end() const
+{
+  return HashSource::HashSourceIterator{this, data_.size()};
+}
+
+HashSource::HashSourceIterator HashSource::cbegin() const
+{
+  return HashSource::HashSourceIterator{this, 0u};
+}
+
+HashSource::HashSourceIterator HashSource::cend() const
+{
+  return HashSource::HashSourceIterator{this, data_.size()};
+}
+
+std::size_t HashSource::getHash(std::size_t index) const
+{
+  return data_[index];
+}
+
+HashSource::HashSourceIterator::~HashSourceIterator() = default;
+
+HashSource::HashSourceIterator::HashSourceIterator(HashSourceIterator const &)     = default;
+HashSource::HashSourceIterator::HashSourceIterator(HashSourceIterator &&) noexcept = default;
+
+HashSource::HashSourceIterator &HashSource::HashSourceIterator::operator=(
+    HashSource::HashSourceIterator const &) = default;
+HashSource::HashSourceIterator &HashSource::HashSourceIterator::operator=(
+    HashSource::HashSourceIterator &&) noexcept = default;
+
+bool HashSource::HashSourceIterator::operator!=(HashSource::HashSourceIterator const &other) const
+{
+  return !(operator==(other));
+}
+
+bool HashSource::HashSourceIterator::operator==(HashSource::HashSourceIterator const &other) const
+{
+  return source_ == other.source_ && hash_index_ == other.hash_index_;
+}
+
+HashSource::HashSourceIterator const HashSource::HashSourceIterator::operator++(int)
+{
+  HashSource::HashSourceIterator temp{source_, hash_index_};
+
+  operator++();
+
+  return temp;
+}
+
+HashSource::HashSourceIterator &HashSource::HashSourceIterator::operator++()
+{
+  ++hash_index_;
+
+  return *this;
+}
+
+std::size_t HashSource::HashSourceIterator::operator*() const
+{
+  return source_->getHash(hash_index_);
+}
+
+HashSource::HashSourceIterator::HashSourceIterator(HashSource const *source, std::size_t index)
+  : hash_index_{index}
+  , source_{source}
+{}
+
+namespace {
+
+class OpenSslHasher
```

Comment:
Do we not already have a similar concept for this in the crypto library?

Reply:
Not really - there's a FNV implementation that I chose to use here, but in general this is quite lightweight. I think there's room for adding some general purpose digest functions to our crypto lib